### PR TITLE
support conditional expression _ ? _ : _

### PR DESCRIPTION
**Pull request by guanqun**

### CODE REVIEW

**Comment by guanqun** [comment]
Do you think "TernaryExpression" is a better name?

**Comment by chriseth** [comment]
Thanks for the pull request! I'll add some comments.

File: test/libsolidity/SolidityEndToEndTest.cpp [hunk 1]
```diff
@@ -86,6 +86,66 @@ BOOST_AUTO_TEST_CASE(exp_operator_const_signed)
...
+ 			}
+		})";
+	compileAndRun(sourceCode);
+	BOOST_CHECK(callContractFunction("f(uint256)", u256(1001)) == toBigEndian(u256(1000)));
+	BOOST_CHECK(callContractFunction("f(uint256)", u256(500)) == toBigEndian(u256(100)));
+	BOOST_CHECK(callContractFunction("f(uint256)", u256(80)) == toBigEndian(u256(50)));
+	BOOST_CHECK(callContractFunction("f(uint256)", u256(40)) == toBigEndian(u256(10)));
+}
+
+/*
```
Comment (chriseth) [inline test/libsolidity/SolidityEndToEndTest.cpp]: Why is this commented out?

File: test/libsolidity/SolidityEndToEndTest.cpp [hunk 2]
```diff
@@ -86,6 +86,66 @@ BOOST_AUTO_TEST_CASE(exp_operator_const_signed)
...
+BOOST_AUTO_TEST_CASE(conditional_expression_as_left_value)
+{
+	char const* sourceCode = R"(
+		contract test {
+			function f(uint x) returns(uint d) {
+				uint y = 1;
+				uint z = 1;
+				(x > 10 ? y : z) = 3;
```
Comment (guanqun) [inline test/libsolidity/SolidityEndToEndTest.cpp]: Because the current compiler doesn't support this line: the syntax of "(xx) = 4", tuple with only one left value.

However, I had another branch to fix this, will send another pull request.

That's why I commented this out.

Comment (chriseth) [inline test/libsolidity/SolidityEndToEndTest.cpp]: ah, ok, I just merged the other PR, so you can add this again.

**Comment by chriseth** [comment]
The fact that `a ? x : y = z` has different semantics in different javascript implementations is worrying. In nodejs, it executes but does not have an effect on either `x` or `y`, in firefox, it assigns to `x` if `a` is true.
We should check the ecmascript 6 specification before finalizing the operator priority and other things in this change.

**Comment by chriseth** [comment]
It seems my initial investigation was wrong. Assignment in the form of `(a ? b : c) = d` is not allowed in JavaScript, but I would not care about that and make sure that it works in Solidity. On the other hand, `a ? b : c = d` should evaluate to the same as `a ? b : (c = d)`.

**Comment by chriseth** [comment]
It would be great if you could also change the documentation:
https://solidity.readthedocs.org/en/latest/control-structures.html#control-structures (it is generated from a file in this source tree in the `docs` directory).

File: libsolidity/analysis/TypeChecker.cpp [hunk 3]
```diff
@@ -737,6 +737,46 @@ void TypeChecker::endVisit(ExpressionStatement const& _statement)
 			typeError(_statement.expression().location(), "Invalid integer constant.");
 }
 
+void TypeChecker::endVisit(Conditional const& _conditional)
```
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp]: Can you use `visit` for this? If yes, then some checks can be shortened to just
`exectType(_conditional.condition(), BoolType());`

File: libsolidity/analysis/TypeChecker.cpp [hunk 4]
```diff
@@ -737,6 +737,46 @@ void TypeChecker::endVisit(ExpressionStatement const& _statement)
...
+			_conditional.location(),
+			"Conditional expression's type " +
+			conditionType->toString() +
+			" doesn't match bool type."
+		);
+
+	TypePointer const& trueType = type(_conditional.trueExpression());
+	TypePointer const& falseType = type(_conditional.falseExpression());
+	// we fake it as an equal operator, but any other comparison operator can work.
+	TypePointer commonType = trueType->binaryOperatorResult(Token::Equal, falseType);
```
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp]: In the case where an lvalue is requested, I fear that you have to require exactly the same type. If an lvalue is not requested, implicit conversions are fine. In this case, I guess it is better to use `Type::commonType()` to find the common type.

File: libsolidity/analysis/TypeChecker.cpp [hunk 5]
```diff
@@ -737,6 +737,46 @@ void TypeChecker::endVisit(ExpressionStatement const& _statement)
...
+		);
+		// even we can't find a common type, we have to set a type here,
+		// otherwise the upper statement will not be able to check the type.
+		commonType = trueType;
+	}
+	_conditional.annotation().type = commonType;
+
+	if (_conditional.annotation().lValueRequested)
+	{
+		requireLValue(_conditional.trueExpression());
```
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp]: This will visit the expression and should thus not be called in the parent's endVisit.

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 6]
```diff
@@ -176,6 +176,20 @@ void ExpressionCompiler::appendStateVariableAccessor(VariableDeclaration const&
...
 
+bool ExpressionCompiler::visit(Conditional const& _condition)
+{
+	CompilerContext::LocationSetter locationSetter(m_context, _condition);
+	_condition.condition().accept(*this);
+	eth::AssemblyItem trueTag = m_context.appendConditionalJump();
+	_condition.falseExpression().accept(*this);
+	eth::AssemblyItem endTag = m_context.appendJumpToNew();
+	m_context << trueTag;
+	m_context.adjustStackOffset(-1);
```
Comment (chriseth) [inline libsolidity/codegen/ExpressionCompiler.cpp]: I think this has to be `m_context.adjustStackOffset(-_condition.annotation().type->sizeOnStack())`.

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 7]
```diff
@@ -176,6 +176,20 @@ void ExpressionCompiler::appendStateVariableAccessor(VariableDeclaration const&
 	m_context.appendJump(eth::AssemblyItem::JumpType::OutOfFunction);
 }
 
+bool ExpressionCompiler::visit(Conditional const& _condition)
+{
+	CompilerContext::LocationSetter locationSetter(m_context, _condition);
+	_condition.condition().accept(*this);
+	eth::AssemblyItem trueTag = m_context.appendConditionalJump();
+	_condition.falseExpression().accept(*this);
```
Comment (chriseth) [inline libsolidity/codegen/ExpressionCompiler.cpp:184]: I think a `utils().convertType(_condition.falseExpression().annotation().type, _condition.annotation().type)` does not hurt (also below).
Please add a test that converts from storage to memory only in one component.

File: libsolidity/parsing/Parser.cpp [hunk 8]
```diff
@@ -930,13 +930,26 @@ ASTPointer<Expression> Parser::parseExpression(
...
+		return nodeFactory.createNode<Assignment>(expression, assignmentOperator, rightHandSide);
+	}
+	else if (m_scanner->currentToken() == Token::Value::Conditional)
+	{
+		m_scanner->next();
+		ASTPointer<Expression> trueExpression = parseExpression();
+		expectToken(Token::Colon);
+		ASTPointer<Expression> falseExpression = parseExpression();
+		ASTNodeFactory nodeFactory(*this, expression);
+		nodeFactory.setEndPositionFromNode(falseExpression); // TODO:
```
Comment (chriseth) [inline libsolidity/parsing/Parser.cpp]: Why the TODO?

File: test/libsolidity/SolidityEndToEndTest.cpp [hunk 9]
```diff
@@ -86,6 +86,66 @@ BOOST_AUTO_TEST_CASE(exp_operator_const_signed)
...
+				return y ** z;
+ 			}
+		})";
+	compileAndRun(sourceCode);
+	BOOST_CHECK(callContractFunction("f(uint256)", u256(20)) == toBigEndian(u256(3)));
+	BOOST_CHECK(callContractFunction("f(uint256)", u256(5)) == toBigEndian(u256(1)));
+}
+*/
+
```
Comment (chriseth) [inline test/libsolidity/SolidityEndToEndTest.cpp:242]: Another test that would be great to have:

```
contract c {
  function f(bool cond, uint v) returns (uint a, unit b) {
    cond ? a = v : b = v;
  }
}
```

File: test/libsolidity/SolidityEndToEndTest.cpp (hunk 2 above)
Comment (guanqun) [inline test/libsolidity/SolidityEndToEndTest.cpp]: Sure. I'll rebase this patch and add it.

File: libsolidity/analysis/TypeChecker.cpp (hunk 3 above)
Comment (guanqun) [inline libsolidity/analysis/TypeChecker.cpp]: OK. I'll use `visit` here.

File: libsolidity/analysis/TypeChecker.cpp (hunk 4 above)
Comment (guanqun) [inline libsolidity/analysis/TypeChecker.cpp]: `Type::commonType()` uses implicit conversion as well, maybe as a left value, we require an exact match?

File: libsolidity/parsing/Parser.cpp (hunk 8 above)
Comment (guanqun) [inline libsolidity/parsing/Parser.cpp]: I forgot to remove this. :)

File: test/libsolidity/SolidityEndToEndTest.cpp (hunk 9 above)
Comment (guanqun) [inline test/libsolidity/SolidityEndToEndTest.cpp:242]: Interesting test case!

Comment (guanqun) [inline test/libsolidity/SolidityEndToEndTest.cpp]: Interesting test case!

File: libsolidity/codegen/ExpressionCompiler.cpp (hunk 6 above)
Comment (guanqun) [inline libsolidity/codegen/ExpressionCompiler.cpp]: Nice catch. Will change it.

File: libsolidity/codegen/ExpressionCompiler.cpp (hunk 7 above)
Comment (guanqun) [inline libsolidity/codegen/ExpressionCompiler.cpp:184]: Oops, I was totally unaware of the necessity of conversion here. Thanks for reminding, I'll add a test for this.

**Comment by guanqun** [comment]
@chriseth I'm disabling conditional expression as left value right now. Because with some digging, it's a little more than I expected, because the current `m_currentLValue` is determined at compile time, but for this conditional expression, it has to be determined at runtime. So we might need to delayed codegen, and what's worse `(true ? a : b)[0]` where `a` is an arry. There has to be some changes in ``LValue.cpp` I'm afraid, I need to put more thoughts on it.

So this PR only supports conditional expression as the right value.

**Comment by chriseth** [comment]
You are right, especially things like the following could be difficult (the type is exactly the same, but the lvalues have to be treated very differently):

```
contract c {
  uint x;
  uint[] y;
  function f(bool s) {
    (s ? x : y.length) = 7;
  }
}
```

JavaScript does not have it, so we do not need it.

File: libsolidity/analysis/TypeChecker.cpp [hunk 10]
```diff
@@ -737,6 +737,48 @@ void TypeChecker::endVisit(ExpressionStatement const& _statement)
...
+	expectType(_conditional.condition(), BoolType());
+
+	_conditional.trueExpression().accept(*this);
+	_conditional.falseExpression().accept(*this);
+
+	TypePointer const& trueType = type(_conditional.trueExpression());
+	TypePointer const& falseType = type(_conditional.falseExpression());
+
+	TypePointer commonType;
+	if (*trueType == *falseType)
```
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp]: You can now directly use `TypePointer commonType = Type::commonType(trueType, falseType);`.

File: test/libsolidity/SolidityEndToEndTest.cpp [hunk 11]
```diff
@@ -86,6 +86,109 @@ BOOST_AUTO_TEST_CASE(exp_operator_const_signed)
...
+	char const* sourceCode = R"(
+		contract test {
+			bytes2[2] data1;
+			function f(bool cond) returns (uint) {
+				bytes2[2] memory x;
+				x[0] = "aa";
+				bytes2[2] memory y;
+				y[0] = "bb";
+
+				data1 = cond ? x : y;
```
Comment (chriseth) [inline test/libsolidity/SolidityEndToEndTest.cpp:155]: Actually I meant something like `x = cond ? y : data1`. As it is now, the conversion just happens after the conditional - which is not a new feature (and does not test the bug that was present in the previous version of the PR).

File: libsolidity/analysis/TypeChecker.cpp (hunk 10 above)
Comment (guanqun) [inline libsolidity/analysis/TypeChecker.cpp]: We can't use `Type::commonType()` because it doesn't handle this case `cond ? 1 : 2`, integers are treated as different types, which should be here.

**Comment by guanqun** [comment]
Can we fix that build? It fails again...

File: libsolidity/analysis/TypeChecker.cpp [hunk 12]
```diff
@@ -737,6 +737,46 @@ void TypeChecker::endVisit(ExpressionStatement const& _statement)
...
+	_conditional.trueExpression().accept(*this);
+	_conditional.falseExpression().accept(*this);
+
+	TypePointer const& trueType = type(_conditional.trueExpression());
+	TypePointer const& falseType = type(_conditional.falseExpression());
+
+	TypePointer commonType = Type::commonType(trueType, falseType);
+	if (!commonType)
+		// we fake it as an equal operator, but any other comparison operator can work.
+		commonType = trueType->binaryOperatorResult(Token::Equal, falseType);
```
Comment (guanqun) [inline libsolidity/analysis/TypeChecker.cpp]: `Type::commonType()` and `binaryOperatorResult()`, maybe we can just add another check for integer const type after `Type::commonType()`, but I think reusing the existing function is more clear.

**Comment by chriseth** [comment]
The build server is busy, it should pick it up soon.

File: libsolidity/analysis/TypeChecker.cpp (hunk 10 above)
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp]: Oh, then use `Type::commonType(trueType->mobileType(), falseType->mobileType())`.

**Comment by guanqun** [comment]
We should add a  button somewhere to trigger the failed build again.

**Comment by VoR0220** [comment]
I tried pulling this and testing it out with just a test contract: 

```
contract c {
    function f() {
        uint x = 12; 
        uint y = 14;

        string S = (x == y) ? "Hello" : "World";

    }
}
```

This throws two errors:

```
testingConditionals.sol:6:14: Error: True expression's type literal_string "Hello" doesn't match false expression's type literal_string "World".
        string S = (x == y) ? "Hello" : "World";
                   ^--------------------------^
testingConditionals.sol:6:3: Error: Type literal_string "Hello" is not implicitly convertible to expected type string storage pointer.
        string S = (x == y) ? "Hello" : "World";
        ^-------------------------------------^
```

I don't think this is how it is supposed to work. Let me know if anybody else is getting this as well.

**Comment by guanqun** [comment]
Hi @VoR0220, thanks for trying out. It's because I'm reusing binary operator comparison (==) as to detect whether the true expression matches the false expression, and it turns out our solidity doesn't support

```
"hello" == "world"
```

What's why you got the error at first place, it seems I need to manually handle these cases instead of reusing existing functions.

**Comment by guanqun** [comment]
I tried more test cases and find another failed case:

```
test.sol:64:17: Error: True expression's type tuple(int_const 1,int_const 2) doesn't match false expression's type tuple(int_const 3,int_const 4).
                true ? (1, 2) : (3, 4);
                ^--------------------^
```

I'll fix these two cases.

**Comment by chriseth** [comment]
As far as I know you do _not_ use the comparison operator to check for equal types, you rather use `Type::commonType` to infer it. The reason it does not work is because you actually first have to convert trueType and falseType to their mobile types and then compute the common type.
This should fix both the literal string and the tuple case.

**Comment by guanqun** [comment]
@chriseth Actually in this implementation, I first check against `Type::commonType` then if it fails, try the binary operator. It means `Type::commonType` doesn't cover the string literal.

I'm trying to fix it by listing the different cases manually here.

File: libsolidity/analysis/TypeChecker.cpp [hunk 13]
```diff
@@ -737,6 +737,46 @@ void TypeChecker::endVisit(ExpressionStatement const& _statement)
...
 }
 
+bool TypeChecker::visit(Conditional const& _conditional)
+{
+	expectType(_conditional.condition(), BoolType());
+
+	_conditional.trueExpression().accept(*this);
+	_conditional.falseExpression().accept(*this);
+
+	TypePointer const& trueType = type(_conditional.trueExpression());
```
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp]: Please convert trueType to its mobile type (and do not store it as a reference in that case).

File: libsolidity/analysis/TypeChecker.cpp [hunk 14]
```diff
@@ -737,6 +737,46 @@ void TypeChecker::endVisit(ExpressionStatement const& _statement)
...
+	TypePointer const& falseType = type(_conditional.falseExpression());
+
+	TypePointer commonType = Type::commonType(trueType, falseType);
+	if (!commonType)
+		// fake it as an equal operator, but any other comparison operator can work.
```
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp]: Please don't do that. This has nothing to do with a comparison operator.

**Comment by chriseth** [comment]
@guanqun did you try the suggestions I made concerning determining the common type?

File: libsolidity/analysis/TypeChecker.cpp (hunk 13 above)
Comment (guanqun) [inline libsolidity/analysis/TypeChecker.cpp]: OK. I'll convert it to mobileType.

**Comment by guanqun** [comment]
```
    TypePointer trueType = type(_conditional.trueExpression())->mobileType();
    TypePointer falseType = type(_conditional.falseExpression())->mobileType();

    TypePointer commonType = Type::commonType(trueType, falseType);
```

@chriseth Now I'm getting mobile type first and then find the common type of these two. Basically it works fine, but still in this case it doesn't work quite nicely:

```
contract C {
    function f(bool cond) returns(bytes32) {
        return cond ? "hello" : "world";
    }
}
```
- The true expression and false expression are both string literals.
- But according to the new implementation, the common type of mobile types is `ArrayType`.
- This `ArrayType` can't be implicitly converted to `bytes32`
- Actually string literal should be converted to `bytes32`.

The error message is:

```
test.sol:3:10: Error: Return argument type string memory is not implicitly convertible to expected type (type of first return variable) bytes32.
        return cond ? "hello" : "world";
               ^----------------------^
```

I'd like to tackle this issue by checking this case specifically (two string literals), and mark the return type of this conditional expression also as a string literal, with the maximum length of these two, adding another field in `StringLiteralType`. 

I'd like to get your idea on this.

**Comment by chriseth** [comment]
The problem is that string literals can be converted into `string`, `bytes` and `bytesXX` (only if they are short enough). As type conversions are only aware of the immediate context (i.e. the other string literal is visible but not the target of the assignment which is one further node away in the expression graph), we only get a `string` or `bytes`.

I'm not sure adding special cases for the ternary operator is a clean way to fix this. One option would be to make `bytesXX` the mobile type for string literals that are short enough and allow `bytesXX` to be implicitly convertible to `bytes` and `string`. This will allow all `bytesXX` types to be convertible to `string` which is probably not a good idea. We might add another flag to `bytesXX` and `bytes` to indicate that the types actually could also be a valid string and allow conversion only in those cases.

In any case, I think that this should be postponed for a different user story.

**Comment by guanqun** [comment]
I'll fix this tonight, tomorrow at most...

**Comment by guanqun** [comment]
I follow your approach to make `bytesXX` the mobile type for string literals and allow it to be implicitly convertible to `bytes` and `string`, but still there are lots of cases failed.

I need to understand more about the relationship between these types before I randomly hack on this.

So for now, I'll close this PR.

**Comment by chriseth** [comment]
I don't think this tiny shortcoming is a reason not to accept this PR. We should add the operator now and worry about these edge cases separately, they are not only relevant for `?:`.

**Comment by VoR0220** [comment]
I agree. We should reopen this.

**Comment by guanqun** [comment]
OK. Now I commented out the string literal test case, and all test cases can pass.

I'll fix this issue eventually after I sorted out all the type conversion rules...

Since if I take the approach to make `bytesXX` the mobile type for string literals, I have to comment more test cases.

**Comment by guanqun** [comment]
Looks like the current `develop` branch has one failure:

```
/webthree-umbrella/solidity/test/libsolidity/SolidityOptimizer.cpp(84): error in "incorrect_storage_access_bug": Computed values do not match.
Non-Optimized: 0000000000000000000000000000000000000000000000000000000000000000
Optimized:     0000000000000000000000000000000000000000000000000000000000000005
```

**Comment by VoR0220** [comment]
I'm hesitant to add the 'looks good' label to it. That optimizer bug is a pretty big problem. I'm going to wait for advice from @chriseth on this one. The optimizer isn't my forte.

**Comment by chriseth** [comment]
The optimizer is implemented in the libethereum submodule. If you update that to latest develop, the error should go away.

**Comment by guanqun** [comment]
Thank you @chriseth @VoR0220 !
